### PR TITLE
feat: Migrate expression evaluation library from `govaluate` to `expr`.

### PR DESCRIPTION
**Pull request by tson1111**
As issue #742 mentioned, we want a more actively-maintained and efficient expression evaluation library. [expr](https://github.com/antonmedv/expr) may be a better choice compared to the current library [govaluate](https://github.com/Knetic/govaluate).

### CODE REVIEW

**Comment by tson1111** [comment]
@hsluoyz My response is as follows.

> 1. Is this a breaking change?
> 2. Can you evaluate the effect of the breaking change?

I cannot guarantee for the prevention of potential risks. And I agree that it's a big change that may introduce some unexpected bugs. All we know is that we passed the unit tests. Maybe we should create a branch for this migration? Or is there anything we can do to test the robustness of this implementation?

> 3. Can you measure the benchmark diff for this PR?

I use benchmark 1.14 for comparison. Please look at the table below. For the fourth column, positive results means `expr` is faster than `gval` while negative means slower.
For most cases, `expr` is slightly faster than `govaluate`. But for some RBAC workloads `expr` has notably worse performance. I believe there's plenty of optimization space for `expr`. But I need more time to try and implement.

```plain text
Benchmark1.14                                           ns/op for expr          ns/op for goval         1 - (expr time/goval time)
BenchmarkCachedRaw-2                                    29.1                    29.9                    2.6755852842809276%
BenchmarkCachedBasicModel-2                             276.0                   283.0                   2.473498233215551%
BenchmarkCachedRBACModel-2                              265.0                   271.0                   2.2140221402214024%
BenchmarkCachedRBACModelSmall-2                         268.0                   282.0                   4.964539007092199%
BenchmarkCachedRBACModelMedium-2                        276.0                   301.0                   8.305647840531563%
BenchmarkCachedRBACModelLarge-2                         326.0                   291.0                   -12.027491408934708%
BenchmarkCachedRBACModelWithResourceRoles-2             270.0                   292.0                   7.534246575342465%
BenchmarkCachedRBACModelWithDomains-2                   285.0                   324.0                   12.037037037037038%
BenchmarkCachedABACModel-2                              9721.0                  9601.0                  -1.2498698052286272%
BenchmarkCachedKeyMatchModel-2                          296.0                   323.0                   8.359133126934978%
BenchmarkCachedRBACModelWithDeny-2                      270.0                   289.0                   6.574394463667815%
BenchmarkCachedPriorityModel-2                          268.0                   293.0                   8.532423208191133%
BenchmarkCachedRBACModelMediumParallel-2                332.0                   278.0                   -19.42446043165468%
BenchmarkHasPolicySmall-2                               1009.0                  1068.0                  5.524344569288388%
BenchmarkHasPolicyMedium-2                              1111.0                  1069.0                  -3.9289055191767943%
BenchmarkHasPolicyLarge-2                               1063.0                  1095.0                  2.9223744292237512%
BenchmarkAddPolicySmall-2                               1049.0                  1042.0                  -0.6717850287907936%
BenchmarkAddPolicyMedium-2                              1351.0                  1446.0                  6.569847856154908%
BenchmarkAddPolicyLarge-2                               2263.0                  2433.0                  6.987258528565562%
BenchmarkRemovePolicySmall-2                            1019.0                  1067.0                  4.498594189315838%
BenchmarkRemovePolicyMedium-2                           1183.0                  1237.0                  4.365400161681492%
BenchmarkRemovePolicyLarge-2                            2299.0                  2306.0                  0.30355594102341854%
BenchmarkRaw-2                                          29.9                    29.9                    0.0%
BenchmarkBasicModel-2                                   32621.0                 18496.0                 -76.36786332179932%
BenchmarkRBACModel-2                                    74661.0                 25495.0                 -192.84565601098257%
BenchmarkRBACModelSmall-2                               1724451.0               197371.0                -773.7104235171327%
BenchmarkRBACModelMedium-2                              18303934.0              2121962.0               -762.5948061275367%
BenchmarkRBACModelLarge-2                               187885208.0             27176345.0              -591.3556918710003%
BenchmarkRBACModelWithResourceRoles-2                   67789.0                 28541.0                 -137.51445289233033%
BenchmarkRBACModelWithDomains-2                         110492.0                37187.0                 -197.12533950036305%
BenchmarkABACModel-2                                    9891.0                  9188.0                  -7.651284283848497%
BenchmarkKeyMatchModel-2                                105220.0                31991.0                 -228.9050045325248%
BenchmarkRBACModelWithDeny-2                            89257.0                 26714.0                 -234.12068578273562%
BenchmarkPriorityModel-2                                24315.0                 19764.0                 -23.026715239829997%
BenchmarkRoleManagerSmall-2                             60476.0                 61829.0                 2.18829351922237%
BenchmarkRoleManagerMedium-2                            638202.0                649090.0                1.6774253185228503
```

> 4. Fix committers

Committers Fixed.


**In summary**, I think it's appropriate to create a branch for `expr` library. It's good for users to have an alternative option and we could enhance the performance and robustness at that branch. Maybe I could draft a proposal about package migration for GSoC or CAS Summer?

**Comment by hsluoyz** [comment]
@tson1111 why is expr even slower in some cases?

image

**Comment by hsluoyz** [comment]
@tson1111 plz make GSoC proposal.

**Comment by tson1111** [comment]
> @tson1111 why is expr even slower in some cases?

I believe it’s due to my implementation. I modified the workflow and this may degrade the performance somehow. I’ll dive into these cases and find out the specific reasons. Any suggestions for profiling tools I could leverage?

> @tson1111 plz make GSoC proposal.

Cool. I’ll try to draft a proposal in the next few days. I would appreciate your feedback. Since the deadline is approaching, I’ll do it ASAP.

File: README.md [hunk 1]
```diff
@@ -111,12 +111,12 @@ Further more, if you are using ABAC,  you can try operator `in` like following i
 ```ini
 # Matchers
 [matchers]
-m = r.obj == p.obj && r.act == p.act || r.obj in ('data2', 'data3')
+m = r.obj == p.obj && r.act == p.act || r.obj in ['data2', 'data3']
```
Comment (Zxilly) [inline README.md:114]: Does this mean we have to change the grammar?
Casbin have multipie language implement. They share the same grammar.

Comment (fr3fou) [reply README.md:114]: I think even before this PR, there were still some differences between the grammars of different implementations.

Comment (hsluoyz) [reply README.md:114]: @Zxilly we will not change the grammar, so this place needs to change.

Comment (hsluoyz) [reply README.md:114]: @fr3fou do you think there are bugs in the different grammars? Please send us github issues!

Comment (tson1111) [reply README.md:114]: @hsluoyz I think the grammar is dependent on the grammar of expression evaluation package. Or do we need to modify `expr` for consistency in `casbin`?

**Comment by andrewsav-bt** [comment]
Is this still going to happen?

**Comment by AndrewSav** [comment]
Is this till going to happen?

**Comment by hsluoyz** [comment]
@AndrewSav can you share your thoughts for preferring expr?

**Comment by AndrewSav** [comment]
@hsluoyz I do not have any, I was just enquiring about status of this work

**Comment by hsluoyz** [comment]
@closetool @tangyang9464 switching to expr may not be bad, but the current state of this PR is not satisfactory and the performance is suffering. It needs more tuning.

**Comment by closetool** [comment]
@hsluoyz Really? It reduces by 700% performance at most, I don't think it is a good idea.
